services/eventservice/src/data_manager.rs

```diff
@@ -174,73 +174,69 @@ impl EventManager {
                     }
+                    (Some(org), None) => self._events_for_organizer(org)?.into_iter(),
+                    (None, Some(cat)) => self._events_for_category(cat)?.into_iter(),
+                    (None, None) => Vec::new().into_iter(),
                 }
-            })
-        } else {
-            let mut ids: Vec<u64> = Vec::new();
-            if organizer.is_some() && category.is_some() {
-                let organizers = self.db.open_tree(&self.organizers_name)?;
-                let org_ids = organizers.get(organizer.unwrap().to_be_bytes())?;
-                let categories = self.db.open_tree(&self.categories_name)?;
-                let cat_ids = categories.get(category.unwrap().as_bytes())?;
-                if org_ids.is_some() && cat_ids.is_some() {
-                    let o_ids: HashSet<u64> = EventManager::_ids_to_vec(&org_ids.unwrap())
-                        .into_iter()
-                        .collect();
-                    let c_ids: HashSet<u64> = EventManager::_ids_to_vec(&cat_ids.unwrap())
-                        .into_iter()
-                        .collect();
-                    let mut res: Vec<u64> = o_ids.intersection(&c_ids).map(|t| t.clone()).collect();
-                    ids.append(&mut res);
-                } else if org_ids.is_some() {
-                    ids.append(&mut EventManager::_ids_to_vec(&org_ids.unwrap()));
-                } else if cat_ids.is_some() {
-                    ids.append(&mut EventManager::_ids_to_vec(&cat_ids.unwrap()));
-                }
-            } else if organizer.is_some() {
-                let organizers = self.db.open_tree(&self.organizers_name)?;
-                let org_ids = organizers.get(organizer.unwrap().to_be_bytes())?;
-                if org_ids.is_some() {
-                    ids.append(&mut EventManager::_ids_to_vec(&org_ids.unwrap()));
-                }
-            } else if category.is_some() {
-                let categories = self.db.open_tree(&self.categories_name)?;
-                let cat_ids = categories.get(category.unwrap().as_bytes())?;
-                if cat_ids.is_some() {
-                    ids.append(&mut EventManager::_ids_to_vec(&cat_ids.unwrap()));
-                }
-            }
-            let events_raw = ids.into_iter().filter_map(|v| events.get(v.to_be_bytes()).ok());
-            events_raw.for_each(|v| {
-                if v.is_some() {
-                    let parsed =
-                        Event::from_json(&String::from_utf8(v.unwrap().to_vec()).unwrap());
-                    if parsed.is_ok() {
-                        return_events.push(parsed.unwrap());
-                    }
-                }
-            });
-        }
-        return_events.sort_by(|a, b| {
+                .filter_map(|v| events.get(v.to_be_bytes()).ok())
+                .filter_map(|v| v)
+                .filter_map(|v| Event::from_json(&String::from_utf8(v.to_vec()).unwrap()).ok())
```

Comment:
can we combine this in one filter?

Like `filter_map(|v|  if events.get(v.to_be_bytes()) && v { Event::from_json(&String::from_utf8(v.to_vec()).unwrap()).ok() }` ?

Reply:
That would be great, but I don't know if it is possible
filter_map just casts `enum Option<T> {Some(T), None}` to T (skipping all None)
I don't know method to unwrap `Option<Option<...Option<T>..>>`

Reply:
You always can try this behavior locally

Reply:
There is even nothing to try

Reply:
It can be done with one `filter_map`, just like this:

```rust
.filter_map(|v| {
   if let Ok(value) = v {
      return value
  }
  None
})
```

Reply:
I suppose filter_map works somehow like this, but I can't see any reason to replace two lines with four

Reply:
Well, you'll have only one pass on the container instead of 3.

Reply:
I see your code is "eating" one option, so I need one more if inside, to have only one filter_map